src/vcpkg/versions.cpp

```diff
@@ -50,93 +49,171 @@ namespace vcpkg::Versions
+    {
+        // Suggested regex by semver.org
+        // ^(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)
+        // (?:-((?:0|[1-9][0-9]*|[0-9]*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9][0-9]*|[0-9]
+        // *[a-zA-Z-][0-9a-zA-Z-]*))*))?
+        // (?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$
 
-        SemanticVersion ret;
+        DotVersion ret;
         ret.original_string = str;
-        ret.version_string = str;
 
-        auto build_found = ret.version_string.find('+');
-        if (build_found != std::string::npos)
+        // (0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)
+        int idx = 0;
+        const char* cur = str.c_str();
+        for (;; ++idx)
         {
-            ret.version_string.resize(build_found);
+            ret.version.push_back(0);
+            cur = parse_skip_number(cur, ret.version.data() + idx);
+            if (!cur || *cur != '.') break;
+            ++cur;
         }
+        if (!cur) goto invalid_semver;
+        ret.version_string.assign(str.c_str(), cur);
+        if (*cur == 0) return ret;
 
-        auto prerelease_found = ret.version_string.find('-');
-        if (prerelease_found != std::string::npos)
+        // pre-release
+        if (*cur == '-')
         {
-            ret.prerelease_string = ret.version_string.substr(prerelease_found + 1);
-            ret.identifiers = Strings::split(ret.prerelease_string, '.');
-            ret.version_string.resize(prerelease_found);
+            ++cur;
+            const char* const start_of_prerelease = cur;
+            for (;;)
+            {
+                const auto start_identifier = cur;
+                cur = skip_prerelease_identifier(cur);
+                if (!cur)
+                {
+                    goto invalid_semver;
+                }
+                ret.identifiers.emplace_back(start_identifier, cur);
+                if (*cur != '.') break;
+                ++cur;
+            }
+            ret.prerelease_string.assign(start_of_prerelease, cur);
         }
+        if (*cur == 0) return ret;
 
-        std::regex version_match("(0|[1-9][0-9]*)(\\.(0|[1-9][0-9]*)){2}");
-        if (!std::regex_match(ret.version_string, version_match))
+        // build
+        if (*cur != '+') goto invalid_semver;
+        ++cur;
+        for (;;)
         {
-            return Strings::format("Error: String `%s` does not follow the required MAJOR.MINOR.PATCH format.",
-                                   ret.version_string);
+            // Require non-empty identifier element
+            if (!Parse::ParserBase::is_alphanumdash(*cur)) goto invalid_semver;
+            ++cur;
+            while (Parse::ParserBase::is_alphanumdash(*cur))
+            {
+                ++cur;
+            }
+            if (*cur == 0) return ret;
+            if (*cur == '.')
+            {
+                ++cur;
+            }
+            else
+            {
+                goto invalid_semver;
```

Comment:
```suggestion
                return nullopt;
```

Ditto the others. I'm not opposed to goto in concept but it seems to not be adding value here.